### PR TITLE
David.benque/scheduling group runner

### PR DESCRIPTION
**Pull request by dbenque**
Building block for scheduling groups:
- indexing: node in a group
- key: interface + key builder based on metadata to have same group than in existing version
- reconciler: ensure that a run is launched for every known node
- groupRunner: run a dedicated go routine (Runner interface) for each group

### CODE REVIEW

File: internal/groups/index.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	cachecr "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	SchedulingGroupIdx = "node:scheduling:group"
+)
+
+// initSchedulingGroupIndexer prepare an index with all the nodes for a given scheduling group
+func initSchedulingGroupIndexer(client client.Client, cache cachecr.Cache, keyGetter GroupKeyGetter) error {
```
Comment (dhenkel92) [inline internal/groups/index.go:17]: We should move this part to the indexer, when my PR is merged

File: internal/groups/key.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	v1 "k8s.io/api/core/v1"
+	"strings"
+)
+
+type GroupKey string
+
+type GroupKeyGetter interface {
+	GetGroupKey(node *v1.Node) GroupKey
```
Comment (dhenkel92) [inline internal/groups/key.go:11]: We could use `metav1.Object` to abstract the node.

File: internal/groups/runner.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func (g *GroupsRunner) runForGroup(key GroupKey) *RunnerInfo {
+	ctx, cancel := context.WithCancel(g.parentContext)
+	r := &RunnerInfo{
+		key:        key,
+		context:    ctx,
+		cancelFunc: cancel,
+	}
+	go func(runInfo *RunnerInfo) {
```
Comment (dhenkel92) [inline internal/groups/runner.go:63]: We should add a test to make sure that the self-cleanup is always working as expected.

File: internal/groups/reconciler.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+	nodeWarmUpDelay time.Duration
+	keyGetter       GroupKeyGetter
+	groupRunner     *GroupsRunner
+}
+
+func NewGroupRegistry(ctx context.Context, kclient client.Client, logger logr.Logger, factory RunnerFactory) *GroupRegistry {
+	return &GroupRegistry{
+		kclient:         kclient,
+		logger:          logger,
```
Comment (dhenkel92) [inline internal/groups/reconciler.go:35]: The logger is not used (yet?).

File: internal/groups/key.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	GroupKeySeparator = "#"
+)
+
+func (g *GroupKeyFromMetadata) GetGroupKey(node *v1.Node) GroupKey {
+	// slice that contains the values that will compose the groupKey
+	var values []string
+
+	// let's tackle the simple case where the user completely override the groupkey
+	if g.groupOverrideAnnotationKey != "" && node.Annotations != nil {
+		if override, ok := node.Annotations[g.groupOverrideAnnotationKey]; ok && override != "" {
```
Comment (dhenkel92) [inline internal/groups/key.go:49]: What is the expected behaviour if the override annotation configuration was set, but the node does not have it?
Shall we fallback to the default mechanism or is it okay to group all the nodes with an empty group name?

**Review by dhenkel92 — COMMENTED** [review]
Looks very good already. 🙂 
It's lot less code than I have expected.

File: internal/groups/index.go (hunk 1 above)
Comment (dbenque) [reply internal/groups/index.go:17]: I am wondering if we should groups of the index initialisations in one place or let the index initialisation live in the package where they are used.
For example this index for scheduling group will not be used out of this package (function is private, and the index name should be). So does it make sense to have the intialisation out of this ?

This being say the Generic code that you have written to GET from the index, can be used in the Runner that will consume the index.

Let's see what make sense once we have all the indexes and the code merged, I am open to both alternatives.

File: internal/groups/runner.go (hunk 3 above)
Comment (dbenque) [reply internal/groups/runner.go:63]: I will do.
I am planing to add a layer for observability, and one of the metrics/primitive will be to count the active runners. With this I can build the test. I am going to amend the PR to do that.

File: internal/groups/reconciler.go (hunk 4 above)
Comment (dbenque) [reply internal/groups/reconciler.go:35]: It is under the groupRunner:
```
groupRunner:     NewGroupsRunner(ctx, factory, logger),
```
and we will probably have more usage directly under the GroupRegistry.
I will probably add a call to `LogForVerboseNode` in the reconcile loop, it is really convenient for debug.

File: internal/groups/key.go (hunk 5 above)
Comment (dbenque) [reply internal/groups/key.go:49]: very good point.
As discussed we will fallback to default behaviour and try to inform user about this misconfiguration

Comment (dbenque) [reply internal/groups/key.go:61]: https://github.com/DataDog/draino/pull/99/commits/87e3548bf8184078a1b6fabe629bf99180492a1f

File: internal/groups/reconciler.go (hunk 4 above)
Comment (dbenque) [reply internal/groups/reconciler.go:40]: using it in https://github.com/DataDog/draino/pull/99/commits/87e3548bf8184078a1b6fabe629bf99180492a1f to log about bad group override value

File: internal/groups/key.go (hunk 2 above)
Comment (dbenque) [reply internal/groups/key.go:11]: Yes it is going to be Node only, using another abstraction would not make sense.

File: internal/groups/runner.go (hunk 3 above)
Comment (dbenque) [reply internal/groups/runner.go:68]: https://github.com/DataDog/draino/pull/99/commits/2a5ade2b8b70a4d4cc5ec7f5e523581e5ab310b2

File: internal/groups/index.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	informer, err := cache.GetInformer(context.Background(), &v1.Node{})
+	if err != nil {
+		return err
+	}
+	return informer.AddIndexers(map[string]cachek.IndexFunc{
+		SchedulingGroupIdx: func(obj interface{}) ([]string, error) {
+			if n, ok := obj.(*v1.Node); ok {
+				return []string{string(keyGetter.GetGroupKey(n))}, nil
+			}
+			return nil, fmt.Errorf("Expecting node object in SchedulingGroupIdx indexer")
```
Comment (dhenkel92) [inline internal/groups/index.go:27]: Do you know when we should use `fmt.Errorf` and when `errors.New`?

File: internal/groups/key.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+			// in that case we completely replace the groups, we remove the default groups.
+			// for example, this allows users to define a kubernetes-cluster wide groups if the default is set to namespace
+			values = strings.Split(override, ",")
+			return GroupKey(strings.Join(values, GroupKeySeparator))
```
Comment (dhenkel92) [inline internal/groups/key.go:65]: [Optional]
```suggestion
			return GroupKey(strings.ReplaceAll(values, ",", GroupKeySeparator))
```

**Review by dhenkel92 — APPROVED** [review]
LGTM 👍

**Pull request merged**